### PR TITLE
Support for pasting images

### LINKED ISSUE #21: Support for pasting images
Extracted from https://github.com/ckeditor/ckeditor5/issues/1184.

We would like to introduce support for images pasted from Word. As for now it only works if image is copied itself (as then it is inserted into clipboard as regular image not Word content) - see https://github.com/ckeditor/ckeditor5/issues/1184#issuecomment-409572471.

When image is a part of the copied content it is not handled correctly. Similar mechanism is already implemented in CKEditor 4 so one can look how it is done there (https://github.com/ckeditor/ckeditor-dev/tree/major/plugins/pastefromword).

### LINKED ISSUE #39: Additional empty paragraph added in a table cell when pasting table from Word
Extracted from #31.

When table is pasted from Word, additional empty paragraph is added on the end of each table cell:

image

when pasted looks like

image

### LINKED ISSUE #40: Empty list item appears in the content after list and paragraph when pasted from Word
I have noticed that on docs sample when entire content is pasted, the empty list item appears before table:

image

and `editor.getData()` returns:

```html
<h2>Recognition of Achievement</h2>
<p>This letter acknowledges the invaluable input <strong>you</strong>, as a member of our <i>Innovation Team</i>,&nbsp;have provided in the “Implement Rich Text Editor” project covering:</p>
<ul>
  <li>Paste from Office feature,</li>
  <li>Tracking changes feature,</li>
  <li>Comments feature.</li>
</ul>
<p>The Management would like to hereby thank you for this great accomplishment that was delivered in a timely fashion, up to the highest company standards, and with great results:</p>
<ul>
  <li>&nbsp; &nbsp; &nbsp; &nbsp;</li>
</ul>
<figure class="table">
...
```

However, the clipboard HTML after PFO processing is:

```html
<h1 align="center" style="margin-bottom:12.0pt;text-align:center"><span lang="EN-US" style="color:black;font-family:">Recognition of Achievement</span></h1>
<p class="MsoNormal" style="text-align:justify;text-justify:inter-ideograph"><span lang="EN-US" style="font-family:">This letter acknowledges the invaluable input <b style="mso-bidi-font-weight:normal">you</b>, as a member of our <i style="mso-bidi-font-style:normal">Innovation Team</i>, have provided in the “Implement Rich Text Editor” project covering:</span></p>
<ul>
  <li class="MsoListParagraphCxSpFirst" style="mso-list:l0 level1 lfo1;text-align:justify;text-indent:-18.0pt;text-justify:inter-ideograph"><span lang="EN-US" style="font-family:Symbol;mso-bidi-font-family:Symbol;mso-fareast-font-family:Symbol"></span><span lang="EN-US" style="font-family:">Paste from Office feature,</span></li>
  <li class="MsoListParagraphCxSpMiddle" style="mso-list:l0 level1 lfo1;text-align:justify;text-indent:-18.0pt;text-justify:inter-ideograph"><span lang="EN-US" style="font-family:Symbol;mso-bidi-font-family:Symbol;mso-fareast-font-family:Symbol"></span><span lang="EN-US" style="font-family:">Tracking changes feature,</span></li>
  <li class="MsoListParagraphCxSpLast" style="mso-list:l0 level1 lfo1;text-align:justify;text-indent:-18.0pt;text-justify:inter-ideograph"><span lang="EN-US" style="font-family:Symbol;mso-bidi-font-family:Symbol;mso-fareast-font-family:Symbol"></span><span lang="EN-US" style="font-family:">Comments feature.</span></li>
</ul>
<p class="MsoNormal" style="text-align:justify;text-justify:inter-ideograph"><span lang="EN-US" style="font-family:">The Management would like to hereby thank you for this great accomplishment that was delivered in a timely fashion, up to the highest company standards, and with great results:</span></p>
<table border="1" cellpadding="0" cellspacing="0" class="MsoTableGrid" style="border:none;border-collapse:collapse;mso-border-alt:solid windowtext .5pt;mso-padding-alt:0cm 5.4pt 0cm 5.4pt;mso-table-layout-alt:fixed;mso-yfti-tbllook:1184" width="597">
 ...
```

so it doesn't seem to be directly caused by PFO processing 🤔

### PR DESCRIPTION
**Pull request by f1ames**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Support for pasting content with images from Word. Closes #21. Closes #39. Closes #40.

---

### Additional information

This PR implements two mechanisms - one for handling images in Chrome, Firefox and Edge and one for handling images in Safari.

### Chrome, Firefox, Edge

When content is pasted from Word, images are represented wit local sources `file://`. In the browser clipboard apart from `text/html` data, there is `text/rtf` data provided which contains entire images encoded as hex strings. These strings are extracted, converted to `base64` and used as new image elements sources e.g ``.

This is a synchronous process so all content manipulation takes place on a view instance so data inserted on paste is in its final form (no transformations are applied after data is rendered in the editor).

### Safari

When content is pasted from Word, images are represented with local blob resource `blob://`. Such resources are then fetched (via `XHR`), converted to `base64` representation and then used as new image elements sources e.g ``.

As fetching blobs is an asynchronous process, first the data with blobs is inserted into the editor (images are already visible) and then when each blob is fetched and converted, images sources are replaced.

**Note**: As this PR was built on top of `t/12` (as it provides whole infrastructure for automatic testing on Safari), it can be merged only after #16 (`t/12` PR) is merged.

### CODE REVIEW

**Comment by f1ames** [comment]
Ready for review 🎉

**Comment by f1ames** [comment]
Just to clarify the previous comment. Moving blob extraction to `UploadImage` plugin means that without the `UploadImage` plugin, images pasted from Word in Safari will be represented as blobs and will not be converted to base64. We have agreed with @Reinmar that this is acceptable solution for now, but may also think about improvements in the future if needed.

**Comment by f1ames** [comment]
This PR contains also some additional [space normalization](https://github.com/ckeditor/ckeditor5-paste-from-office/blob/0f1ea89fa3c7f0a297556ab4d49eaeccd6befd5b/src/filters/space.js#L23) which apparently solves #39 and #40, I have added to `Closes` commit message.

**Comment by Reinmar** [comment]
Haha :D While checking the code I thought... shouldn't it fix the issue that @f1ames just reported? :D

File: src/filters/space.js [hunk 1]
```diff
@@ -11,14 +11,16 @@
  * Replaces last space preceding elements closing tag with `&nbsp;`. Such operation prevents spaces from being removed
  * during further DOM/View processing (see especially {@link module:engine/view/domconverter~DomConverter#_processDataFromDomText}).
  * This method also takes into account Word specific `<o:p></o:p>` empty tags.
+ * Additionally multiline sequences of spaces and new lines between tags are removed.
```
Comment (Reinmar) [inline src/filters/space.js]: Mention what bugs it fixed.

**Comment by f1ames** [comment]
image

😄 👍

**Comment by Reinmar** [comment]
I've done some testing and it works as expected. I'm going to have a quick look at the code still (especially in https://github.com/ckeditor/ckeditor5-image/pull/248) but we should be able to merge this very soon.

File: src/filters/image.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+	const imageElementsMatcher = new ViewMatcher( {
+		name: 'img'
+	} );
+
+	const imgs = [];
+
+	for ( const value of range ) {
+		if ( imageElementsMatcher.match( value.item ) ) {
+			if ( value.item.getAttribute( 'src' ).indexOf( 'file://' ) === 0 ) {
```
Comment (Reinmar) [inline src/filters/image.js]: `startsWith()`

File: src/filters/image.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	const regexPictureHeader = /{\\pict[\s\S]+?\\bliptag-?\d+(\\blipupi-?\d+)?({\\\*\\blipuid\s?[\da-fA-F]+)?[\s}]*?/;
+	const regexPicture = new RegExp( '(?:(' + regexPictureHeader.source + '))([\\da-fA-F\\s]+)\\}', 'g' );
+	const images = rtfData.match( regexPicture );
+	const result = [];
+
+	if ( images ) {
+		for ( const image of images ) {
+			let imageType = false;
+
+			if ( image.indexOf( '\\pngblip' ) !== -1 ) {
```
Comment (Reinmar) [inline src/filters/image.js]: `includes()`

File: src/filters/image.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+				imageType = 'image/png';
+			} else if ( image.indexOf( '\\jpegblip' ) !== -1 ) {
```
Comment (Reinmar) [inline src/filters/image.js]: `includes()`

File: src/filters/utils.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+/* globals btoa */
+
+/**
+ * Converts given HEX string to base64 representation.
+ *
+ * @param {String} hexString The HEX string to be converted.
+ * @returns {String} Base64 representation of a given HEX string.
+ */
+export function convertHexToBase64( hexString ) {
```
Comment (Reinmar) [inline src/filters/utils.js]: Does this util have to be public? It's used in one place. If you want to test it, export it from the file where it's used as an underscored, protected util.

**Pull request merged**